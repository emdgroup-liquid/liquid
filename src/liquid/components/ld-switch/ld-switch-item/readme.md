---
eleventyNavigation:
  key: Switch Item
  parent: Switch
layout: layout.njk
title: Switch Item
permalink: components/ld-switch/ld-switch-item/
---

# ld-switch-item

The `ld-switch-item` component is a subcomponent for `ld-switch`.

Please refer to the [`ld-switch` documentation](components/ld-switch) for usage examples.


<!-- Auto Generated Below -->


## Properties

| Property       | Attribute       | Description                                              | Type               | Default     |
| -------------- | --------------- | -------------------------------------------------------- | ------------------ | ----------- |
| `ariaDisabled` | `aria-disabled` | Alternative disabled state that keeps element focusable  | `string`           | `undefined` |
| `checked`      | `checked`       | Indicates whether the switch item is selected.           | `boolean`          | `false`     |
| `disabled`     | `disabled`      | Disabled state of the switch item.                       | `boolean`          | `undefined` |
| `key`          | `key`           | for tracking the node's identity when working with lists | `string \| number` | `undefined` |
| `ref`          | `ref`           | reference to component                                   | `any`              | `undefined` |
| `value`        | `value`         | The input value.                                         | `string`           | `undefined` |


## Methods

### `focusInner() => Promise<void>`

Sets focus on the switch item.

#### Returns

Type: `Promise<void>`




## Shadow Parts

| Part              | Description                                   |
| ----------------- | --------------------------------------------- |
| `"content"`       | content container element                     |
| `"input"`         | the form input element                        |
| `"label"`         | text label container containing the main slot |
| `"label-element"` | wrapping label element                        |


----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
