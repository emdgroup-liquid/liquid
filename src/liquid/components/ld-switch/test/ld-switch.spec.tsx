import { newSpecPage } from '@stencil/core/testing'
import { LdSwitch } from '../ld-switch'
import { LdSwitchItem } from '../ld-switch-item/ld-switch-item'
import { LdIcon } from '../../ld-icon/ld-icon'
import '../../../utils/mutationObserver'

const components = [LdSwitch, LdSwitchItem, LdIcon]

describe('ld-switch', () => {
  it('renders', async () => {
    const page = await newSpecPage({
      components,
      html: `
        <ld-switch legend="Dress" name="dress">
          <ld-switch-item value="werewolf" checked>Werewolf</ld-switch-item>
          <ld-switch-item value="zombie">Zombie</ld-switch-item>
          <ld-switch-item value="mummy" disabled>Mummy</ld-switch-item>
          <ld-switch-item value="vampire" aria-disabled="true">Vampire</ld-switch-item>
        </ld-switch>
      `,
    })
    expect(page.root).toMatchSnapshot()
  })

  it('with icons only', async () => {
    const page = await newSpecPage({
      components,
      html: `
        <ld-switch legend="Dress" name="dress">
          <ld-switch-item value="werewolf" checked>
            <ld-icon slot="icon-start" name="placeholder" aria-label="Werewolf"></ld-icon>
          </ld-switch-item>
          <ld-switch-item value="zombie">
            <ld-icon slot="icon-start" name="placeholder" aria-label="Zombie"></ld-icon>
          </ld-switch-item>
        </ld-switch>
      `,
    })
    expect(page.root).toMatchSnapshot()
  })

  describe('css classes', () => {
    it('brand-color', async () => {
      const page = await newSpecPage({
        components,
        html: `
          <ld-switch name="dress" brand-color>
            <ld-switch-item value="werewolf" checked>Werewolf</ld-switch-item>
            <ld-switch-item value="zombie">Zombie</ld-switch-item>
          </ld-switch>
        `,
      })
      expect(page.root).toHaveClass('ld-switch--brand-color')
    })

    it('fit-content', async () => {
      const page = await newSpecPage({
        components,
        html: `
          <ld-switch name="dress" fit-content>
            <ld-switch-item value="werewolf" checked>Werewolf</ld-switch-item>
            <ld-switch-item value="zombie">Zombie</ld-switch-item>
          </ld-switch>
        `,
      })
      expect(page.root).toHaveClass('ld-switch--fit-content')
    })

    it('size', async () => {
      const page = await newSpecPage({
        components,
        html: `
          <ld-switch name="dress" size="sm">
            <ld-switch-item value="werewolf" checked>Werewolf</ld-switch-item>
            <ld-switch-item value="zombie">Zombie</ld-switch-item>
          </ld-switch>
        `,
      })
      expect(page.root).toHaveClass('ld-switch--sm')
    })
  })

  describe('mouse interactions', () => {
    it('changes item via click', async () => {
      const page = await newSpecPage({
        components,
        html: `
          <ld-switch name="dress">
            <ld-switch-item value="werewolf" checked>Werewolf</ld-switch-item>
            <ld-switch-item value="zombie">Zombie</ld-switch-item>
            <ld-switch-item value="mummy">Mummy</ld-switch-item>
          </ld-switch>
        `,
      })
      const ldSwitch = page.root
      const ldSwitchItems = ldSwitch.querySelectorAll('ld-switch-item')
      expect(ldSwitchItems.length).toEqual(3)

      const ldSwitchItem = ldSwitchItems[2]
      const switchItemInput = ldSwitchItem.shadowRoot.querySelector('input')

      expect(switchItemInput.getAttribute('tabindex')).toEqual('-1')
      expect(switchItemInput.checked).toBe(false)

      ldSwitchItem.click()
      await page.waitForChanges()

      expect(switchItemInput).not.toHaveAttribute('tabindex')
      expect(switchItemInput.checked).toBe(true)
    })

    it('does not change item via click on disabled item', async () => {
      const page = await newSpecPage({
        components,
        html: `
          <ld-switch name="dress">
            <ld-switch-item value="werewolf" checked>Werewolf</ld-switch-item>
            <ld-switch-item value="zombie" disabled>Zombie</ld-switch-item>
            <ld-switch-item value="mummy">Mummy</ld-switch-item>
          </ld-switch>
        `,
      })
      const ldSwitch = page.root
      const ldSwitchItems = ldSwitch.querySelectorAll('ld-switch-item')
      expect(ldSwitchItems.length).toEqual(3)

      const ldSwitchItem = ldSwitchItems[1]
      const switchItemInput = ldSwitchItem.shadowRoot.querySelector('input')

      expect(switchItemInput.getAttribute('tabindex')).toEqual('-1')
      expect(switchItemInput.checked).toBe(false)
      expect(ldSwitchItem.disabled).toBe(true)
      expect(switchItemInput.disabled).toBe(true)

      ldSwitchItem.dispatchEvent(new Event('click'))
      await page.waitForChanges()

      expect(switchItemInput.getAttribute('tabindex')).toEqual('-1')
      expect(switchItemInput.checked).toBe(false)
    })

    it('does not select item if switch is disabled', async () => {
      const page = await newSpecPage({
        components,
        html: `
          <ld-switch name="dress" disabled>
            <ld-switch-item value="werewolf" checked>Werewolf</ld-switch-item>
            <ld-switch-item value="zombie">Zombie</ld-switch-item>
            <ld-switch-item value="mummy">Mummy</ld-switch-item>
          </ld-switch>
        `,
      })
      const ldSwitch = page.root
      const ldSwitchItems = ldSwitch.querySelectorAll('ld-switch-item')
      expect(ldSwitchItems.length).toEqual(3)

      const ldSwitchItem = ldSwitchItems[1]
      const switchItemInput = ldSwitchItem.shadowRoot.querySelector('input')

      expect(switchItemInput.getAttribute('tabindex')).toEqual('-1')
      expect(switchItemInput.checked).toBe(false)
      expect(ldSwitchItem.disabled).toBe(true)
      expect(switchItemInput.disabled).toBe(true)

      ldSwitchItem.click()
      await page.waitForChanges()

      expect(switchItemInput.getAttribute('tabindex')).toEqual('-1')
      expect(switchItemInput.checked).toBe(false)

      const disabledLdSwitchItems = Array.from(ldSwitchItems).filter(
        (item) => item.disabled
      )
      expect(disabledLdSwitchItems.length).toEqual(3)
    })

    it('does not select item if switch is readonly', async () => {
      const page = await newSpecPage({
        components,
        html: `
          <ld-switch name="dress" readonly>
            <ld-switch-item value="werewolf" checked>Werewolf</ld-switch-item>
            <ld-switch-item value="zombie">Zombie</ld-switch-item>
            <ld-switch-item value="mummy">Mummy</ld-switch-item>
          </ld-switch>
        `,
      })
      const ldSwitch = page.root
      const ldSwitchItems = ldSwitch.querySelectorAll('ld-switch-item')
      expect(ldSwitchItems.length).toEqual(3)

      const ldSwitchItem = ldSwitchItems[1]
      const switchItemInput = ldSwitchItem.shadowRoot.querySelector('input')

      expect(switchItemInput.getAttribute('tabindex')).toEqual('-1')
      expect(switchItemInput.checked).toBe(false)
      expect(ldSwitchItem.readonly).toBe(true)

      ldSwitchItem.click()
      await page.waitForChanges()

      expect(switchItemInput.getAttribute('tabindex')).toEqual('-1')
      expect(switchItemInput.checked).toBe(false)

      const readonlyLdSwitchItems = Array.from(ldSwitchItems).filter(
        (item) => item.readonly
      )
      expect(readonlyLdSwitchItems.length).toEqual(3)
    })

    it('does not change item via click on aria-disabled item', async () => {
      const page = await newSpecPage({
        components,
        html: `
          <ld-switch name="dress">
            <ld-switch-item value="werewolf" checked>Werewolf</ld-switch-item>
            <ld-switch-item value="zombie" aria-disabled="true">Zombie</ld-switch-item>
            <ld-switch-item value="mummy">Mummy</ld-switch-item>
          </ld-switch>
        `,
      })
      const ldSwitch = page.root
      const ldSwitchItems = ldSwitch.querySelectorAll('ld-switch-item')
      expect(ldSwitchItems.length).toEqual(3)

      const ldSwitchItem = ldSwitchItems[1]
      const switchItemInput = ldSwitchItem.shadowRoot.querySelector('input')

      expect(switchItemInput.getAttribute('tabindex')).toEqual('-1')
      expect(switchItemInput.checked).toBe(false)
      expect(switchItemInput.getAttribute('aria-disabled')).toEqual('true')

      ldSwitchItem.dispatchEvent(new Event('click'))
      await page.waitForChanges()

      expect(switchItemInput.getAttribute('tabindex')).toEqual('-1')
      expect(switchItemInput.checked).toBe(false)
    })

    it('does not select item while switch is aria-disabled', async () => {
      const page = await newSpecPage({
        components,
        html: `
          <ld-switch name="dress" aria-disabled="true">
            <ld-switch-item value="werewolf" checked>Werewolf</ld-switch-item>
            <ld-switch-item value="zombie">Zombie</ld-switch-item>
            <ld-switch-item value="mummy">Mummy</ld-switch-item>
          </ld-switch>
        `,
      })
      const ldSwitch = page.root
      const ldSwitchItems = ldSwitch.querySelectorAll('ld-switch-item')
      expect(ldSwitchItems.length).toEqual(3)

      const ldSwitchItem = ldSwitchItems[1]
      const switchItemInput = ldSwitchItem.shadowRoot.querySelector('input')

      expect(switchItemInput.getAttribute('tabindex')).toEqual('-1')
      expect(switchItemInput.checked).toBe(false)
      expect(ldSwitch.getAttribute('aria-disabled')).toEqual('true')

      ldSwitchItem.click()
      await page.waitForChanges()

      expect(switchItemInput.getAttribute('tabindex')).toEqual('-1')
      expect(switchItemInput.checked).toBe(false)

      const disabledLdSwitchItems = Array.from(ldSwitchItems).filter(
        (item) => item.ariaDisabled
      )
      expect(disabledLdSwitchItems.length).toEqual(3)
    })
  })

  describe('keyboard interactions', () => {
    it('changes swichItem via arrow keys', async () => {
      const page = await newSpecPage({
        components,
        html: `
          <ld-switch name="dress">
            <ld-switch-item value="werewolf" checked>Werewolf</ld-switch-item>
            <ld-switch-item value="zombie">Zombie</ld-switch-item>
            <ld-switch-item value="mummy">Mummy</ld-switch-item>
          </ld-switch>
        `,
      })
      const ldSwitch = page.root
      const ldSwitchItems = ldSwitch.querySelectorAll('ld-switch-item')
      expect(ldSwitchItems.length).toEqual(3)

      const [ldSwitchItem0, ldSwitchItem1, ldSwitchItem2] =
        Array.from(ldSwitchItems)

      const item0 = ldSwitchItem0.shadowRoot.querySelector('input')
      const item1 = ldSwitchItem1.shadowRoot.querySelector('input')
      const item2 = ldSwitchItem2.shadowRoot.querySelector('input')

      expect(item0).not.toHaveAttribute('tabindex')
      expect(item1.getAttribute('tabindex')).toEqual('-1')
      expect(item2.getAttribute('tabindex')).toEqual('-1')

      item0.focus = jest.fn()
      item1.focus = jest.fn()
      item2.focus = jest.fn()

      item0.dispatchEvent(
        new KeyboardEvent('keydown', { key: 'ArrowRight', bubbles: true })
      )
      await page.waitForChanges()

      expect(item0.getAttribute('tabindex')).toEqual('-1')
      expect(item0.checked).toBe(false)

      expect(item1).not.toHaveAttribute('tabindex')
      expect(item1.checked).toBe(true)

      expect(item2.getAttribute('tabindex')).toEqual('-1')
      expect(item2.checked).toBe(false)

      item1.dispatchEvent(
        new KeyboardEvent('keydown', { key: 'ArrowDown', bubbles: true })
      )
      await page.waitForChanges()

      expect(item0.getAttribute('tabindex')).toEqual('-1')
      expect(item0.checked).toBe(false)

      expect(item1.getAttribute('tabindex')).toEqual('-1')
      expect(item1.checked).toBe(false)

      expect(item2).not.toHaveAttribute('tabindex')
      expect(item2.checked).toBe(true)

      item2.dispatchEvent(
        new KeyboardEvent('keydown', { key: 'ArrowLeft', bubbles: true })
      )
      await page.waitForChanges()

      expect(item0.getAttribute('tabindex')).toEqual('-1')
      expect(item0.checked).toBe(false)

      expect(item1).not.toHaveAttribute('tabindex')
      expect(item1.checked).toBe(true)

      expect(item2.getAttribute('tabindex')).toEqual('-1')
      expect(item2.checked).toBe(false)

      item1.dispatchEvent(
        new KeyboardEvent('keydown', { key: 'ArrowUp', bubbles: true })
      )
      await page.waitForChanges()

      expect(item0).not.toHaveAttribute('tabindex')
      expect(item0.checked).toBe(true)

      expect(item1.getAttribute('tabindex')).toEqual('-1')
      expect(item1.checked).toBe(false)

      expect(item2.getAttribute('tabindex')).toEqual('-1')
      expect(item2.checked).toBe(false)
    })
  })

  describe('events', () => {
    it('emits ldswitchchange event', async () => {
      const page = await newSpecPage({
        components,
        html: `
          <ld-switch name="dress">
            <ld-switch-item value="werewolf" checked>Werewolf</ld-switch-item>
            <ld-switch-item value="zombie">Zombie</ld-switch-item>
            <ld-switch-item value="mummy" disabled>Mummy</ld-switch-item>
          </ld-switch>
        `,
      })

      const ldSwitch = page.root
      const ldSwitchItems = ldSwitch.querySelectorAll('ld-switch-item')
      const ldSwitchItem = ldSwitchItems[1]

      const handleSwitchChange = jest.fn()
      ldSwitch.addEventListener('ldswitchchange', handleSwitchChange)

      ldSwitchItem.click()
      await page.waitForChanges()

      expect(handleSwitchChange).toHaveBeenCalled()
    })
  })

  describe('focus', () => {
    it('sets inner focus on checked item', async () => {
      const page = await newSpecPage({
        components,
        html: `
          <ld-switch name="dress">
            <ld-switch-item value="werewolf" checked>Werewolf</ld-switch-item>
            <ld-switch-item value="zombie">Zombie</ld-switch-item>
            <ld-switch-item value="mummy">Mummy</ld-switch-item>
          </ld-switch>
        `,
      })
      const ldSwitch = page.root
      const ldSwitchItems = ldSwitch.querySelectorAll('ld-switch-item')
      const ldSwitchItem = ldSwitchItems[0]
      const input = ldSwitchItem.shadowRoot.querySelector('input')

      input.focus = jest.fn()
      await ldSwitch.focusInner()

      expect(input.focus).toHaveBeenCalled()
    })

    it('updates tabindices on focus and focusout', async () => {
      const page = await newSpecPage({
        components,
        html: `
          <ld-switch name="dress">
            <ld-switch-item value="werewolf" checked>Werewolf</ld-switch-item>
            <ld-switch-item value="zombie">Zombie</ld-switch-item>
            <ld-switch-item value="mummy">Mummy</ld-switch-item>
          </ld-switch>
        `,
      })
      const ldSwitch = page.root
      const ldSwitchItems = ldSwitch.querySelectorAll('ld-switch-item')
      const ldSwitchItem0 = ldSwitchItems[0]
      const ldSwitchItem1 = ldSwitchItems[1]
      const ldSwitchItem2 = ldSwitchItems[2]
      const input0 = ldSwitchItem0.shadowRoot.querySelector('input')
      const input1 = ldSwitchItem1.shadowRoot.querySelector('input')
      const input2 = ldSwitchItem2.shadowRoot.querySelector('input')

      expect(ldSwitch.tabIndex).toEqual(0)
      expect(input0.tabIndex).toEqual(-1)
      expect(input1.tabIndex).toEqual(-1)
      expect(input2.tabIndex).toEqual(-1)

      input0.focus = jest.fn()
      ldSwitch.dispatchEvent(new FocusEvent('focus'))

      await page.waitForChanges()

      expect(input0.focus).toHaveBeenCalled()

      input0.dispatchEvent(new FocusEvent('focus'))

      await page.waitForChanges()

      expect(ldSwitch.tabIndex).toEqual(-1)
      expect(input0.tabIndex).toEqual(-1)
      expect(input1.tabIndex).toEqual(-1)
      expect(input2.tabIndex).toEqual(-1)

      ldSwitch.dispatchEvent(new FocusEvent('focusout'))

      await page.waitForChanges()

      expect(ldSwitch.tabIndex).toEqual(0)
      expect(input0.tabIndex).toEqual(-1)
      expect(input1.tabIndex).toEqual(-1)
      expect(input2.tabIndex).toEqual(-1)
    })

    it('updates tabindices on focus and focusout with custom tabindex', async () => {
      const page = await newSpecPage({
        components,
        html: `
          <ld-switch name="dress" ld-tabindex="2">
            <ld-switch-item value="werewolf" checked>Werewolf</ld-switch-item>
            <ld-switch-item value="zombie">Zombie</ld-switch-item>
            <ld-switch-item value="mummy">Mummy</ld-switch-item>
          </ld-switch>
        `,
      })
      const ldSwitch = page.root
      const ldSwitchItems = ldSwitch.querySelectorAll('ld-switch-item')
      const ldSwitchItem0 = ldSwitchItems[0]
      const ldSwitchItem1 = ldSwitchItems[1]
      const ldSwitchItem2 = ldSwitchItems[2]
      const input0 = ldSwitchItem0.shadowRoot.querySelector('input')
      const input1 = ldSwitchItem1.shadowRoot.querySelector('input')
      const input2 = ldSwitchItem2.shadowRoot.querySelector('input')

      expect(ldSwitch.tabIndex).toEqual(2)
      expect(input0.tabIndex).toEqual(2)
      expect(input1.tabIndex).toEqual(-1)
      expect(input2.tabIndex).toEqual(-1)

      input0.focus = jest.fn()
      ldSwitch.dispatchEvent(new FocusEvent('focus'))

      await page.waitForChanges()

      expect(input0.focus).toHaveBeenCalled()

      input0.dispatchEvent(new FocusEvent('focus'))

      await page.waitForChanges()

      expect(ldSwitch.tabIndex).toEqual(-1)
      expect(input0.tabIndex).toEqual(2)
      expect(input1.tabIndex).toEqual(-1)
      expect(input2.tabIndex).toEqual(-1)

      ldSwitch.dispatchEvent(new FocusEvent('focusout'))

      await page.waitForChanges()

      expect(ldSwitch.tabIndex).toEqual(2)
      expect(input0.tabIndex).toEqual(2)
      expect(input1.tabIndex).toEqual(-1)
      expect(input2.tabIndex).toEqual(-1)
    })

    it('sets inner focus on first item if none is checked', async () => {
      const page = await newSpecPage({
        components,
        html: `
          <ld-switch name="dress">
            <ld-switch-item value="werewolf">Werewolf</ld-switch-item>
            <ld-switch-item value="zombie">Zombie</ld-switch-item>
            <ld-switch-item value="mummy">Mummy</ld-switch-item>
          </ld-switch>
        `,
      })
      const ldSwitch = page.root
      const ldSwitchItems = ldSwitch.querySelectorAll('ld-switch-item')
      const ldSwitchItem = ldSwitchItems[0]
      const input = ldSwitchItem.shadowRoot.querySelector('input')

      input.focus = jest.fn()
      await ldSwitch.focusInner()

      expect(input.focus).toHaveBeenCalled()
    })
  })

  it('creates hidden input field, if inside a form', async () => {
    const page = await newSpecPage({
      components,
      html: `
        <form>
          <ld-switch name="dress">
            <ld-switch-item value="werewolf" checked>Werewolf</ld-switch-item>
            <ld-switch-item value="zombie">Zombie</ld-switch-item>
            <ld-switch-item value="mummy">Mummy</ld-switch-item>
          </ld-switch>
        </form>
      `,
    })
    expect(page.root).toMatchSnapshot()
  })

  it('uses hidden input field with referenced form', async () => {
    const page = await newSpecPage({
      components,
      html: `
        <ld-switch name="dress" form="yolo">
          <ld-switch-item value="werewolf" checked>Werewolf</ld-switch-item>
          <ld-switch-item value="zombie">Zombie</ld-switch-item>
          <ld-switch-item value="mummy">Mummy</ld-switch-item>
        </ld-switch>
      `,
    })
    expect(page.root).toMatchSnapshot()
  })

  it('is required', async () => {
    const page = await newSpecPage({
      components,
      html: `
        <ld-switch name="dress" required>
          <ld-switch-item value="werewolf" checked>Werewolf</ld-switch-item>
          <ld-switch-item value="zombie">Zombie</ld-switch-item>
          <ld-switch-item value="mummy" disabled>Mummy</ld-switch-item>
          <ld-switch-item value="vampire" aria-disabled="true">Vampire</ld-switch-item>
        </ld-switch>
      `,
    })
    const ldSwitch = page.root
    const ldSwitchItems = ldSwitch.querySelectorAll('ld-switch-item')
    expect(ldSwitchItems.length).toEqual(4)

    const requiredInputs = Array.from(ldSwitchItems).filter(
      (ldSwitchItem) => ldSwitchItem.required
    )
    expect(requiredInputs.length).toEqual(4)
  })

  it('removes hidden input field if name prop is removed', async () => {
    const page = await newSpecPage({
      components,
      html: `
        <ld-switch name="dress" form="my-form">
          <ld-switch-item value="werewolf" checked>Werewolf</ld-switch-item>
          <ld-switch-item value="zombie">Zombie</ld-switch-item>
        </ld-switch>
      `,
    })
    const ldSwitch = page.root
    const ldSwitchItems = ldSwitch.querySelectorAll('ld-switch-item')
    expect(ldSwitchItems[0].querySelector('input')).toHaveProperty(
      'name',
      'dress'
    )

    ldSwitch.removeAttribute('name')
    await page.waitForChanges()

    expect(ldSwitchItems[0].querySelectorAll('input').length).toEqual(0)
  })

  it('removes value prop on hidden input', async () => {
    const page = await newSpecPage({
      components,
      html: `
        <ld-switch name="dress" form="my-form">
          <ld-switch-item value="werewolf" checked>Werewolf</ld-switch-item>
          <ld-switch-item value="zombie">Zombie</ld-switch-item>
        </ld-switch>
      `,
    })
    const ldSwitch = page.root
    const ldSwitchItems = ldSwitch.querySelectorAll('ld-switch-item')
    expect(ldSwitchItems[0].querySelector('input')).toHaveProperty(
      'value',
      'werewolf'
    )

    ldSwitchItems[0].removeAttribute('value')
    await page.waitForChanges()

    expect(ldSwitchItems[0].querySelector('input')).toHaveProperty('value', '')
  })

  it('removes form prop on hidden input', async () => {
    const page = await newSpecPage({
      components,
      html: `
        <form>
          <ld-switch name="dress" form="my-form">
            <ld-switch-item value="werewolf" checked>Werewolf</ld-switch-item>
            <ld-switch-item value="zombie">Zombie</ld-switch-item>
          </ld-switch>
        </form>
      `,
    })
    const ldSwitch = page.body.querySelector('ld-switch')
    const ldSwitchItems = ldSwitch.querySelectorAll('ld-switch-item')
    expect(ldSwitchItems[0].querySelector('input')).toHaveProperty(
      'form',
      'my-form'
    )

    ldSwitch.removeAttribute('form')
    await page.waitForChanges()

    expect(ldSwitchItems[0].querySelector('input')).toHaveProperty('form', '')
  })

  it('removes hidden input field with removal of form prop when there is no outer form', async () => {
    const page = await newSpecPage({
      components,
      html: `
        <ld-switch name="dress" form="my-form">
          <ld-switch-item value="werewolf" checked>Werewolf</ld-switch-item>
          <ld-switch-item value="zombie">Zombie</ld-switch-item>
        </ld-switch>
      `,
    })
    const ldSwitch = page.root
    const ldSwitchItems = ldSwitch.querySelectorAll('ld-switch-item')
    expect(ldSwitchItems[0].querySelector('input')).toHaveProperty(
      'form',
      'my-form'
    )

    ldSwitch.removeAttribute('form')
    await page.waitForChanges()

    expect(ldSwitchItems[0].querySelectorAll('input').length).toEqual(0)
  })
})
