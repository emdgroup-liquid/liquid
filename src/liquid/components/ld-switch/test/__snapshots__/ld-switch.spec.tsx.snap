// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ld-switch creates hidden input field, if inside a form 1`] = `
<ld-switch class="ld-switch" name="dress" tabindex="0">
  <mock:shadow-root>
    <fieldset part="fieldset">
      <slot></slot>
    </fieldset>
  </mock:shadow-root>
  <ld-switch-item checked="" class="ld-switch-item ld-switch-item--has-label" value="werewolf">
    <mock:shadow-root>
      <label part="label-element">
        <input checked="" part="input focusable" type="radio" value="werewolf">
        <span class="ld-switch-item__content" part="content">
          <slot name="icon-start"></slot>
          <span class="ld-switch-item__label" part="label">
            <slot></slot>
          </span>
          <slot name="icon-end"></slot>
        </span>
      </label>
    </mock:shadow-root>
    Werewolf
    <input checked="" name="dress" type="radio" value="werewolf" style="visibility: hidden; position: absolute; pointer-events: none;">
  </ld-switch-item>
  <ld-switch-item class="ld-switch-item ld-switch-item--has-label" value="zombie">
    <mock:shadow-root>
      <label part="label-element">
        <input part="input focusable" tabindex="-1" type="radio" value="zombie">
        <span class="ld-switch-item__content" part="content">
          <slot name="icon-start"></slot>
          <span class="ld-switch-item__label" part="label">
            <slot></slot>
          </span>
          <slot name="icon-end"></slot>
        </span>
      </label>
    </mock:shadow-root>
    Zombie
    <input name="dress" type="radio" value="zombie" style="visibility: hidden; position: absolute; pointer-events: none;">
  </ld-switch-item>
  <ld-switch-item class="ld-switch-item ld-switch-item--has-label" value="mummy">
    <mock:shadow-root>
      <label part="label-element">
        <input part="input focusable" tabindex="-1" type="radio" value="mummy">
        <span class="ld-switch-item__content" part="content">
          <slot name="icon-start"></slot>
          <span class="ld-switch-item__label" part="label">
            <slot></slot>
          </span>
          <slot name="icon-end"></slot>
        </span>
      </label>
    </mock:shadow-root>
    Mummy
    <input name="dress" type="radio" value="mummy" style="visibility: hidden; position: absolute; pointer-events: none;">
  </ld-switch-item>
</ld-switch>
`;

exports[`ld-switch renders 1`] = `
<ld-switch class="ld-switch" legend="Dress" name="dress" tabindex="0">
  <mock:shadow-root>
    <fieldset part="fieldset">
      <legend part="legend">
        Dress
      </legend>
      <slot></slot>
    </fieldset>
  </mock:shadow-root>
  <ld-switch-item checked="" class="ld-switch-item ld-switch-item--has-label" value="werewolf">
    <mock:shadow-root>
      <label part="label-element">
        <input checked="" part="input focusable" type="radio" value="werewolf">
        <span class="ld-switch-item__content" part="content">
          <slot name="icon-start"></slot>
          <span class="ld-switch-item__label" part="label">
            <slot></slot>
          </span>
          <slot name="icon-end"></slot>
        </span>
      </label>
    </mock:shadow-root>
    Werewolf
  </ld-switch-item>
  <ld-switch-item class="ld-switch-item ld-switch-item--has-label" value="zombie">
    <mock:shadow-root>
      <label part="label-element">
        <input part="input focusable" tabindex="-1" type="radio" value="zombie">
        <span class="ld-switch-item__content" part="content">
          <slot name="icon-start"></slot>
          <span class="ld-switch-item__label" part="label">
            <slot></slot>
          </span>
          <slot name="icon-end"></slot>
        </span>
      </label>
    </mock:shadow-root>
    Zombie
  </ld-switch-item>
  <ld-switch-item class="ld-switch-item ld-switch-item--has-label" disabled="" value="mummy">
    <mock:shadow-root>
      <label part="label-element">
        <input disabled="" part="input focusable" tabindex="-1" type="radio" value="mummy">
        <span class="ld-switch-item__content" part="content">
          <slot name="icon-start"></slot>
          <span class="ld-switch-item__label" part="label">
            <slot></slot>
          </span>
          <slot name="icon-end"></slot>
        </span>
      </label>
    </mock:shadow-root>
    Mummy
  </ld-switch-item>
  <ld-switch-item aria-disabled="true" class="ld-switch-item ld-switch-item--has-label" value="vampire">
    <mock:shadow-root>
      <label part="label-element">
        <input aria-disabled="true" part="input focusable" tabindex="-1" type="radio" value="vampire">
        <span class="ld-switch-item__content" part="content">
          <slot name="icon-start"></slot>
          <span class="ld-switch-item__label" part="label">
            <slot></slot>
          </span>
          <slot name="icon-end"></slot>
        </span>
      </label>
    </mock:shadow-root>
    Vampire
  </ld-switch-item>
</ld-switch>
`;

exports[`ld-switch uses hidden input field with referenced form 1`] = `
<ld-switch class="ld-switch" form="yolo" name="dress" tabindex="0">
  <mock:shadow-root>
    <fieldset part="fieldset">
      <slot></slot>
    </fieldset>
  </mock:shadow-root>
  <ld-switch-item checked="" class="ld-switch-item ld-switch-item--has-label" value="werewolf">
    <mock:shadow-root>
      <label part="label-element">
        <input checked="" part="input focusable" type="radio" value="werewolf">
        <span class="ld-switch-item__content" part="content">
          <slot name="icon-start"></slot>
          <span class="ld-switch-item__label" part="label">
            <slot></slot>
          </span>
          <slot name="icon-end"></slot>
        </span>
      </label>
    </mock:shadow-root>
    Werewolf
    <input checked="" form="yolo" name="dress" type="radio" value="werewolf" style="visibility: hidden; position: absolute; pointer-events: none;">
  </ld-switch-item>
  <ld-switch-item class="ld-switch-item ld-switch-item--has-label" value="zombie">
    <mock:shadow-root>
      <label part="label-element">
        <input part="input focusable" tabindex="-1" type="radio" value="zombie">
        <span class="ld-switch-item__content" part="content">
          <slot name="icon-start"></slot>
          <span class="ld-switch-item__label" part="label">
            <slot></slot>
          </span>
          <slot name="icon-end"></slot>
        </span>
      </label>
    </mock:shadow-root>
    Zombie
    <input form="yolo" name="dress" type="radio" value="zombie" style="visibility: hidden; position: absolute; pointer-events: none;">
  </ld-switch-item>
  <ld-switch-item class="ld-switch-item ld-switch-item--has-label" value="mummy">
    <mock:shadow-root>
      <label part="label-element">
        <input part="input focusable" tabindex="-1" type="radio" value="mummy">
        <span class="ld-switch-item__content" part="content">
          <slot name="icon-start"></slot>
          <span class="ld-switch-item__label" part="label">
            <slot></slot>
          </span>
          <slot name="icon-end"></slot>
        </span>
      </label>
    </mock:shadow-root>
    Mummy
    <input form="yolo" name="dress" type="radio" value="mummy" style="visibility: hidden; position: absolute; pointer-events: none;">
  </ld-switch-item>
</ld-switch>
`;

exports[`ld-switch with icons only 1`] = `
<ld-switch class="ld-switch" legend="Dress" name="dress" tabindex="0">
  <mock:shadow-root>
    <fieldset part="fieldset">
      <legend part="legend">
        Dress
      </legend>
      <slot></slot>
    </fieldset>
  </mock:shadow-root>
  <ld-switch-item checked="" class="ld-switch-item" value="werewolf">
    <mock:shadow-root>
      <label part="label-element">
        <input checked="" part="input focusable" type="radio" value="werewolf">
        <span class="ld-switch-item__content" part="content">
          <slot name="icon-start"></slot>
          <span class="ld-switch-item__label" part="label">
            <slot></slot>
          </span>
          <slot name="icon-end"></slot>
        </span>
      </label>
    </mock:shadow-root>
    <ld-icon aria-label="Werewolf" class="ld-icon" name="placeholder" role="presentation" slot="icon-start">
      <mock:shadow-root>
        Not Found
      </mock:shadow-root>
    </ld-icon>
  </ld-switch-item>
  <ld-switch-item class="ld-switch-item" value="zombie">
    <mock:shadow-root>
      <label part="label-element">
        <input part="input focusable" tabindex="-1" type="radio" value="zombie">
        <span class="ld-switch-item__content" part="content">
          <slot name="icon-start"></slot>
          <span class="ld-switch-item__label" part="label">
            <slot></slot>
          </span>
          <slot name="icon-end"></slot>
        </span>
      </label>
    </mock:shadow-root>
    <ld-icon aria-label="Zombie" class="ld-icon" name="placeholder" role="presentation" slot="icon-start">
      <mock:shadow-root>
        Not Found
      </mock:shadow-root>
    </ld-icon>
  </ld-switch-item>
</ld-switch>
`;
